From 3eb53f3b201da4f931f21a59b518add326dbe10c Mon Sep 17 00:00:00 2001
From: sai surya teja <sssteja@habana.ai>
Date: Tue, 21 Mar 2023 08:27:11 +0200
Subject: [PATCH] [SW-93113] hl: preboot: fix out of bound write issue in crc8 library

This patch fixes an issue pointed by synopsis static analysis tool by
avoiding an out of bound write to table array in crc8_populate_msb
by using a modulo operator to make sure the writes to this array are
within the array length limits.

"47755 Out-of-bounds write
This could cause an immediate crash or incorrect computations.
In crc8_populate_msb: Out-of-bounds write to a buffer (CWE-119)"

Change-Id: If077a49fc28a4e18bfa241cf854ccfaa1c216766
---

diff --git a/lib/preboot/crc8.c b/lib/preboot/crc8.c
index ffcab61..cfd9869 100644
--- a/lib/preboot/crc8.c
+++ b/lib/preboot/crc8.c
@@ -26,8 +26,14 @@
 
 	for (i = 1; i < CRC8_TABLE_SIZE; i *= 2) {
 		t = (t << 1) ^ (t & msbit ? polynomial : 0);
-		for (j = 0; j < i; j++)
-			table[i+j] = table[j] ^ t;
+		for (j = 0; j < i; j++) {
+			/*
+			 * Using (i+j) % CRC8_TABLE_SIZE to ensure no out of bound write.
+			 * This is a deviation from standard library as we noticed index wrap around
+			 * not happening with arc_gnu_2019 compiler.
+			 */
+			table[(i+j) % CRC8_TABLE_SIZE] = table[j] ^ t;
+		}
 	}
 }
 
